src/mapping/mod.rs

```diff
@@ -247,6 +248,67 @@ impl Function for MergeFn {
 
 //------------------------------------------------------------------------------
 
+/// Represents the different log levels that can be used by LogFn
+#[derive(Debug, Clone, Copy)]
+pub(in crate::mapping) enum LogLevel {
+    Trace,
+    Debug,
+    Info,
+    Warn,
+    Error,
+}
+
+impl TryFrom<&str> for LogLevel {
+    type Error = String;
+
+    fn try_from(level: &str) -> Result<Self> {
+        match level {
+            "trace" => Ok(Self::Trace),
+            "debug" => Ok(Self::Debug),
+            "info" => Ok(Self::Info),
+            "warn" => Ok(Self::Warn),
+            "error" => Ok(Self::Error),
+            _ => Err("invalid log level".to_string()),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(in crate::mapping) struct LogFn {
+    msg: Box<dyn query::Function>,
+    level: Option<LogLevel>,
+}
+
+impl LogFn {
+    pub(in crate::mapping) fn new(msg: Box<dyn query::Function>, level: Option<LogLevel>) -> Self {
+        Self { msg, level }
+    }
+}
+
+impl Function for LogFn {
+    fn apply(&self, target: &mut Event) -> Result<()> {
+        let msg = match self.msg.execute(target)? {
+            QueryValue::Value(value) => value,
+            _ => return Err("can only log Value parameters".to_string()),
```

Comment:
It'd be odd for a user to want to do this, but I feel like we could just log `QueryValue::Regex()` as its string representation to avoid this special casing. We could implement `Display` for `QueryValue`.

Reply:
With the work @JeanMertz is doing for #4695 the way this is structured is going to change.

It's a good point, I think we need to decide when moving this function over whether logging a regex should be a compile error or should we log it. Regexes are only allowed as parameters for functions that can actually make use of them (`split` and `replace` so far). Like you say it would be odd to want to do this, so I think probably just not allowing it in the compiler would make sense.

Reply:
:+1: if we can determine it at compile time (config load time), I support that.

If we can't do it until runtime, I guess I'd prefer to just allow logging of regexes as their string representation than raise an error.

Reply:
Yeah, it would be at load time.